projector/binding.go

```diff
@@ -42,69 +52,354 @@ func New(mappingSource MappingSource) ServiceBindingProjector {
 	if err != nil {
 		return err
 	}
-	if err := i.unproject(ctx, binding, mpt); err != nil {
-		return err
-	}
+	p.unproject(binding, mpt)
 	return mpt.WriteToWorkload(ctx)
 }
 
-func (i *serviceBindingProjector) project(ctx context.Context, binding *servicebindingv1alpha3.ServiceBinding, mpt *metaPodTemplate) error {
+func (p *serviceBindingProjector) project(binding *servicebindingv1alpha3.ServiceBinding, mpt *metaPodTemplate) {
 	// rather than attempt to merge an existing binding, unproject it
-	if err := i.unproject(ctx, binding, mpt); err != nil {
-		return err
+	p.unproject(binding, mpt)
+
+	if p.secretName(binding) == "" {
+		// no secret to bind
+		return
 	}
+	p.projectVolume(binding, mpt)
+	for i := range mpt.Containers {
+		p.projectContainer(binding, mpt, &mpt.Containers[i])
+	}
+}
 
+func (p *serviceBindingProjector) unproject(binding *servicebindingv1alpha3.ServiceBinding, mpt *metaPodTemplate) {
+	p.unprojectVolume(binding, mpt)
 	for i := range mpt.Containers {
-		c := &mpt.Containers[i]
-		// TODO skip container if not allowed
-
-		serviceBindingRoot := ""
-		for _, e := range c.Env {
-			if e.Name == ServiceBindingRootEnv {
-				serviceBindingRoot = e.Value
-				break
-			}
+		p.unprojectContainer(binding, mpt, &mpt.Containers[i])
+	}
+
+	// cleanup annotations
+	delete(mpt.Annotations, p.secretAnnotationName(binding))
+	delete(mpt.Annotations, p.typeAnnotationName(binding))
+	delete(mpt.Annotations, p.providerAnnotationName(binding))
+}
+
+func (p *serviceBindingProjector) projectVolume(binding *servicebindingv1alpha3.ServiceBinding, mpt *metaPodTemplate) {
+	volume := corev1.Volume{
+		Name: p.volumeName(binding),
+		VolumeSource: corev1.VolumeSource{
+			Projected: &corev1.ProjectedVolumeSource{
+				Sources: []corev1.VolumeProjection{
+					{
+						Secret: &corev1.SecretProjection{
+							LocalObjectReference: corev1.LocalObjectReference{
+								Name: p.secretAnnotation(binding, mpt),
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	if binding.Spec.Type != "" {
+		volume.VolumeSource.Projected.Sources = append(volume.VolumeSource.Projected.Sources,
+			corev1.VolumeProjection{
+				DownwardAPI: &corev1.DownwardAPIProjection{
+					Items: []corev1.DownwardAPIVolumeFile{
+						{
+							Path: "type",
+							FieldRef: &corev1.ObjectFieldSelector{
+								FieldPath: fmt.Sprintf("metadata.annotations['%s']", p.typeAnnotation(binding, mpt)),
+							},
+						},
+					},
+				},
+			},
+		)
+	}
+	if binding.Spec.Provider != "" {
+		volume.VolumeSource.Projected.Sources = append(volume.VolumeSource.Projected.Sources,
+			corev1.VolumeProjection{
+				DownwardAPI: &corev1.DownwardAPIProjection{
+					Items: []corev1.DownwardAPIVolumeFile{
+						{
+							Path: "provider",
+							FieldRef: &corev1.ObjectFieldSelector{
+								FieldPath: fmt.Sprintf("metadata.annotations['%s']", p.providerAnnotation(binding, mpt)),
+							},
+						},
+					},
+				},
+			},
+		)
+	}
```

Comment:
I thought on adding first annotations for `type` and `provider` that overwrites values in the secret and then to iterate here over those, instead of checking if `Provider` or `Type` are defined in the binding.